### PR TITLE
Added support for `architectures` parameter in @EnabledOnOs and @DisabledOnOs

### LINKED ISSUE #2848: Support architecture in @EnabledOnOs and @DisabledOnOs
Hi team,

In Spring Boot we [had the issue](https://github.com/spring-projects/spring-boot/issues/30082) that some of our tests don't run on Linux when using an AArch64 architecture. I then added a [custom annotation](https://github.com/spring-projects/spring-boot/blob/24a68fc5953f7de1c485b70692e65002782f4d70/spring-boot-project/spring-boot-tools/spring-boot-test-support/src/main/java/org/springframework/boot/testsupport/junit/DisabledOnOs.java) and the [condition](https://github.com/spring-projects/spring-boot/blob/24a68fc5953f7de1c485b70692e65002782f4d70/spring-boot-project/spring-boot-tools/spring-boot-test-support/src/main/java/org/springframework/boot/testsupport/junit/DisabledOnOsCondition.java) to disable the tests on Linux AArch64. This could be something that JUnit Jupiter provides out of the box. 

What do you think?

Another note: This could lead to making the annotations repeatable, as maybe a test doesn't work on Linux aarch64 and on MacOS x86, which you could not do without repeating the annotation.

## Deliverables

- [ ] Introduce `architecture` attribute in the `@EnabledOnOs` and `@DisabledOnOs` annotations.

### PR DESCRIPTION
**Pull request by michaelfmnk**
## Overview

Resolves #2848 

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Review by juliette-derancourt — CHANGES_REQUESTED** [review]
Thank you @michaelfmnk for the PR! :+1: 
I left a few comments and questions :slightly_smiling_face:

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.9.0-M1.adoc [hunk 1]
```diff
@@ -102,6 +102,7 @@ on GitHub.
 * `JAVA_19` has been added to the `JRE` enum for use with JRE-based execution conditions.
 * `@MethodSource` factory methods can now accept arguments resolved by registered
   `ParameterResolver` extensions.
+* Added support for `architectures` parameter in @EnabledOnOs and @DisabledOnOs
```
[the remark this answers is not in the document]
Comment (michaelfmnk) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.9.0-M1.adoc]: Sure! Moved to 5.9.0-M2 file

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOsCondition.java [hunk 2]
```diff
@@ -32,9 +31,21 @@ class DisabledOnOsCondition extends BooleanExecutionCondition<DisabledOnOs> {
...
+	}
+
+	private boolean isEnabledBasedOnArchitecture(DisabledOnOs annotation) {
+		String arch = getArchitecture();
+		return Arrays.stream(annotation.architectures()).noneMatch(arch::equalsIgnoreCase);
+	}
+
+	private boolean isEnabledBasedOnOs(DisabledOnOs annotation) {
 		OS[] operatingSystems = annotation.value();
-		Preconditions.condition(operatingSystems.length > 0, "You must declare at least one OS in @DisabledOnOs");
```
[the remark this answers is not in the document]
Comment (michaelfmnk) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOsCondition.java:36]: Yeah, makes sense. Added preconditions for DisabledOnOs & EnabledOnOs

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java [hunk 3]
```diff
@@ -23,19 +22,34 @@
...
+
+	private boolean isEnabledBasedOnArchitecture(EnabledOnOs annotation) {
+		String[] architectures = annotation.architectures();
+		String arch = getArchitecture();
+		return Arrays.stream(architectures).anyMatch(arch::equalsIgnoreCase);
+	}
+
+	private boolean isEnabledBasedOnOs(EnabledOnOs annotation) {
 		OS[] operatingSystems = annotation.value();
-		Preconditions.condition(operatingSystems.length > 0, "You must declare at least one OS in @EnabledOnOs");
```
[the remark this answers is not in the document]
Comment (michaelfmnk) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java:37]: done

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java [hunk 4]
```diff
@@ -23,19 +22,34 @@
...
+	static final String DISABLED_ON_CURRENT_OS = String.format("Disabled on operating system: %s (%s)",
+		System.getProperty("os.name"), System.getProperty("os.arch"));
 
 	EnabledOnOsCondition() {
 		super(EnabledOnOs.class, ENABLED_ON_CURRENT_OS, DISABLED_ON_CURRENT_OS, EnabledOnOs::disabledReason);
 	}
 
 	@Override
 	boolean isEnabled(EnabledOnOs annotation) {
+		return isEnabledBasedOnOs(annotation) || isEnabledBasedOnArchitecture(annotation);
```
[the remark this answers is not in the document]
Comment (michaelfmnk) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java:40]: I just revisited the #2848 issue. I think you're right. In one of the solutions ("Solution 3"), marcphilipp uses AND condition for enabling a test. I'll update this.

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/api/condition/DisabledOnOsConditionTests.java [hunk 5]
```diff
@@ -157,14 +157,32 @@ void other() {
...
+
+	/**
+	 * @see DisabledOnOsIntegrationTests#architectureAarch64()
+	 */
+	@Test
+	void architectureAarch64() {
+		evaluateCondition();
+		assertDisabledOnCurrentOsIf(onArchitecture("aarch64"));
+	}
+
```
[the remark this answers is not in the document]
Comment (michaelfmnk) [reply junit-jupiter-engine/src/test/java/org/junit/jupiter/api/condition/DisabledOnOsConditionTests.java:180]: Sure. I added more tests for combinations of Mac, Windows, Linux with x86_64 and aarch64 parameters.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (michaelfmnk) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java]: done

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOsCondition.java [hunk 6]
```diff
@@ -32,9 +31,21 @@ class DisabledOnOsCondition extends BooleanExecutionCondition<DisabledOnOs> {
 
 	@Override
 	boolean isEnabled(DisabledOnOs annotation) {
+		return isEnabledBasedOnOs(annotation) && isEnabledBasedOnArchitecture(annotation);
```
[the remark this answers is not in the document]
Comment (michaelfmnk) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOsCondition.java]: done

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java [hunk 7]
```diff
@@ -23,19 +22,34 @@
  */
 class EnabledOnOsCondition extends BooleanExecutionCondition<EnabledOnOs> {
 
-	static final String ENABLED_ON_CURRENT_OS = "Enabled on operating system: " + System.getProperty("os.name");
+	static final String ENABLED_ON_CURRENT_OS = String.format("Enabled on operating system: %s (%s)",
```
[the remark this answers is not in the document]
Comment (michaelfmnk) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java:26]: It would be much better! I just couldn't find a way to pass enabled/disabled reason to BooleanExecutionCondition as values that are calculated based on isEnabled call. 

I think I can refactor BooleanExecutionCondition to accept enabled/disabled reason as functions, but then isEnabled calculation would have to be called twice: actual isEnabled call & inside reason calculation. 

The other option would be to refactor isEnabled method signature so it returns a composite object with true/false & reason inside. 
```java
abstract EnablingResult isEnabled(A annotation);
```

However, it requires much more refactoring and will affect all BooleanExecutionCondition children. What do you think?

[the remark this answers is not in the document]
Comment (sbrannen) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java:26]: `EnablingResult` sounds a lot like `ConditionEvaluationResult`, kinda like reinventing the wheel.

> However, it requires much more refactoring and will affect all BooleanExecutionCondition children. What do you think?

I don't think I'd make that change in a way that affects all `BooleanExecutionCondition` subclasses.

Instead, if you go that route, I'd probably introduce two variants of the abstract base class. For example, the unaffected subclasses could continue to extend `BooleanExecutionCondition`, and the OS variants could extend a new `AbstractOsBasedExecutionCondition` (or similar). And I probably wouldn't try too hard to avoid code duplication in the new abstract base class. If you can avoid some code duplication then good, but if you can't easily avoid code duplication I wouldn't worry about it too much. 

Disclaimer: this is just my personal opinion. 😉

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.9.0-M2.adoc [hunk 8]
```diff
@@ -24,6 +24,7 @@ on GitHub.
 ==== New Features and Improvements
 
 * New public factory method to instantiate an `ExecutionRequest`.
+* Added support for `architectures` parameter in @EnabledOnOs and @DisabledOnOs
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.9.0-M2.adoc:27]:
```suggestion
* Support for hardware architecture in `@EnabledOnOs` and `@DisabledOnOs` via new `architectures` attributes
```

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOs.java [hunk 9]
```diff
@@ -24,10 +24,11 @@
 /**
  * {@code @DisabledOnOs} is used to signal that the annotated test class or
  * test method is <em>disabled</em> on one or more specified
- * {@linkplain #value operating systems}.
+ * {@linkplain #value operating systems} or on one or more specified
+ * {@linkplain #architectures architectures}
  *
  * <p>When applied at the class level, all test methods within that class
- * will be disabled on the same specified operating systems.
+ * will be disabled on the same specified operating systems or architectures.
```
Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOs.java:31]: We need to clearly document AND vs. OR semantics, potentially in a dedicated paragraph.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOs.java [hunk 10]
```diff
@@ -24,10 +24,11 @@
 /**
  * {@code @EnabledOnOs} is used to signal that the annotated test class or
  * test method is only <em>enabled</em> on one or more specified
- * {@linkplain #value operating systems}.
+ * {@linkplain #value operating systems} or one or more specified
+ * {@linkplain #architectures architectures}.
  *
  * <p>When applied at the class level, all test methods within that class
- * will be enabled on the same specified operating systems.
+ * will be enabled on the same specified operating systems or architectures.
```
Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOs.java:31]: We need to clearly document AND vs. OR semantics, potentially in a dedicated paragraph.

**Review by sbrannen — CHANGES_REQUESTED** [review]
This is looking good.

Please address the minor issues I've raised, and we'll see if anyone else in the team (@juliette-derancourt?) has further input.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOsCondition.java [hunk 11]
```diff
@@ -24,17 +22,47 @@
...
 		OS[] operatingSystems = annotation.value();
-		Preconditions.condition(operatingSystems.length > 0, "You must declare at least one OS in @DisabledOnOs");
+		if (operatingSystems.length == 0) {
+			return false;
+		}
+
 		return Arrays.stream(operatingSystems).noneMatch(OS::isCurrentOs);
 	}
 
+	protected String getArchitecture() {
```
[the remark this answers is not in the document]
Comment (michaelfmnk) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOsCondition.java]: Meant to do that. Thanks for noticing. Just updated PR with this change.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java [hunk 12]
```diff
@@ -21,21 +22,46 @@
...
+	private boolean isEnabledBasedOnOs(EnabledOnOs annotation) {
 		OS[] operatingSystems = annotation.value();
-		Preconditions.condition(operatingSystems.length > 0, "You must declare at least one OS in @EnabledOnOs");
+		if (operatingSystems.length == 0) {
+			return true;
+		}
 		return Arrays.stream(operatingSystems).anyMatch(OS::isCurrentOs);
 	}
 
+	protected String getArchitecture() {
```
[the remark this answers is not in the document]
Comment (michaelfmnk) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java]: Just updated PR with this change

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOs.java (hunk 9 above)
Comment (michaelfmnk) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOs.java:31]: how about "will be enabled/disabled on the same specified combination of operating systems and architectures."?

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (juliette-derancourt) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java]: Hmm, I didn't realize we couldn't easily read the content of the annotation from here.

Now that I know, I don't think it's worth it to add all that complexity just to change the message. I would be ok keeping the original one, if y'all agree?

If we want to highlight the fact that OS and architecture are not necessarily enabled/disabled jointly, the message could be more like `Enabled on operating system %s and architecture %s`? But the usage of "and" might be even more confusing... 🤔

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOs.java (hunk 9 above)
Comment (juliette-derancourt) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/DisabledOnOs.java:31]: > We need to clearly document AND vs. OR semantics, potentially in a dedicated paragraph.

I might even be worth a clarification in the user guide?

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (marcphilipp) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/EnabledOnOsCondition.java]: We briefly discussed this in the team call and concluded that we think the clearer message justifies the added complexity.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/AbstractOsBasedExecutionCondition.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+import static org.junit.jupiter.api.extension.ConditionEvaluationResult.enabled;
+import static org.junit.platform.commons.util.AnnotationUtils.findAnnotation;
+
+import java.lang.annotation.Annotation;
+
+import org.junit.jupiter.api.extension.ConditionEvaluationResult;
+import org.junit.jupiter.api.extension.ExecutionCondition;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+abstract class AbstractOsBasedExecutionCondition<A extends Annotation> implements ExecutionCondition {
```
[the remark this answers is not in the document]
Comment (marcphilipp) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/condition/AbstractOsBasedExecutionCondition.java:22]: Done in bd2f68a220ea46b6e3715ff1859ed3046740c48d.

**Review by marcphilipp — APPROVED** [review]

**Comment by marcphilipp** [comment]
The test failures are unrelated (caused by the new `ThreadMode.SEPARATE_THREAD`) and I'll fix them after merging this PR.

**Comment by marcphilipp** [comment]
@michaelfmnk Thanks a lot for your PR! I made some small changes but you certainly did the bulk of the work.

**Comment by sbrannen** [comment]
> @michaelfmnk Thanks a lot for your PR! I made some small changes but you certainly did the bulk of the work.

👍
